tee-worker/omni-executor/intent/executors/cross-chain/src/cross_chain.rs

```diff
@@ -361,126 +389,270 @@ impl<BinanceClient: BinanceApi, SolanaClient: SolanaClientTrait>
-			// - `cummulativeQuoteQty`` shows the total amount of the quote-asset spent
-			let mut trade_success = false;
-			let mut bnb_acquired = "".to_string();
-			loop {
-				let trade_order = SpotTradingApi::new(self.binance_api.as_ref())
-					.get_order(&trade_symbol, Some(binance_order.order_id), None, None)
-					.await
-					.map_err(|_| {
-						error!("Failed to get binance order");
-					})?;
-
-				match trade_order.status {
-					BinanceOrderStatus::FILLED => {
-						info!("Binance order filled");
-						bnb_acquired = match trade_order.side {
-							BinanceOrderSide::BUY => trade_order.executed_qty,
-							BinanceOrderSide::SELL => trade_order.cummulative_quote_qty,
-						};
-						trade_success = true;
-						break;
-					},
-					BinanceOrderStatus::CANCELED => {
-						error!("Binance order canceled");
-					},
-					BinanceOrderStatus::REJECTED => {
-						error!("Binance order rejected");
-					},
-					BinanceOrderStatus::EXPIRED => {
-						error!("Binance order expired");
-					},
-					BinanceOrderStatus::EXPIRED_IN_MATCH => {
-						error!("Binance order expired in matching");
+	async fn estimate_payout(
+		&self,
+		from_asset: ChainAsset,
+		from_amount: Decimal,
+	) -> Result<String, ()> {
+		// TODO: create an util function to convert ChainAsset to binance names
+		// and create constants for SOL, USDC, USDT, etc
+		let (_from_network_name, binance_coin_name, _) = match from_asset {
+			ChainAsset::Solana(ref token) => {
+				let (asset, token_address) = match token {
+					SolanaToken::Native => ("SOL", ""),
+					SolanaToken::SPL(mint_address) => {
+						let mint_address_string = mint_address.as_ref().to_base58();
+						match mint_address_string.as_str() {
+							SOLANA_USDC_MINT_ADDRESS => ("USDC", SOLANA_USDC_MINT_ADDRESS),
+							SOLANA_USDT_MINT_ADDRESS => ("USDT", SOLANA_USDT_MINT_ADDRESS),
+							_ => {
+								error!("Unsupported SPL token: {:?}", mint_address);
+								return Err(());
+							},
+						}
 					},
-					_ => {
-						debug!("Binance order status: {:?}", trade_order.status);
+				};
+				("SOL".to_string(), asset.to_string(), token_address.to_string())
+			},
+			ChainAsset::Ethereum(..) => {
+				error!("Unsupported from_asset: {:?}", from_asset);
+				return Err(());
+			},
+		};
+
+		let (trade_symbol, _) = match binance_coin_name.as_str() {
+			"USDC" => ("BNBUSDC".to_string(), BinanceOrderSide::BUY),
+			"USDT" => ("BNBUSDT".to_string(), BinanceOrderSide::BUY),
+			"SOL" => ("SOLBNB".to_string(), BinanceOrderSide::SELL),
+			_ => {
+				error!("Unsupported asset: {:?}", binance_coin_name);
+				return Err(());
+			},
+		};
+
+		// init `payout_amount` with estimated-amount-to-receive
+		let payout_amount =
+			estimate_bnb_amount(&self.binance_api, &trade_symbol, &binance_coin_name, from_amount)
+				.await?;
+		Ok(payout_amount)
+	}
+
+	// TODO: move `should_wait_for_deposit_confirm` to provider config
+	async fn do_binance_trade(
```

Comment:
This comments seems to be outdated. `should_wait_for_deposit_confirm` is not used in this function

Reply:
I decided to kept it https://github.com/litentry/heima/blob/c95859a569531175a04d1f18c6d21cc9477ba41a/tee-worker/omni-executor/intent/executors/cross-chain/src/cross_chain.rs#L158

Reply:
The TODO is ok in the other functions, what I mean is that `should_wait_for_deposit_confirm` is not used in `do_binance_trade`, so the comment is confusing

Reply:
But it's moved in this PR (there is only one place where this todo exists), do you have any idea which place would be better to keep it ?

Reply:
I think it would be a good idea to put it where `should_wait_for_deposit_confirm` is defined. What I don't know is when/how this value will change and when are we doing the binance trades